examples/fit_with_bounds.py

```diff
@@ -18,78 +18,92 @@
-    variable_names = tuple([str(j) for j in variable_symbols])
-    parameter_symbols = tuple(parameter_symbols)
-    parameter_names = tuple([str(j) for j in parameter_symbols])
-    logger.info(
-        strm(
-            "all symbols are",
-            all_symbols,
-            "axis names are",
-            axis_names,
-            "variable names are",
-            variable_names,
-            "parameter names are",
-            parameter_names,
+        Returns
+        =======
+        pars: lmfit.Parameters
+        fn: function
+            the fit function
+        """
+        # {{{ decide which symbols are parameters vs. variables
+        all_symbols = expr.atoms(sp.Symbol)
+        axis_names = set([sp.Symbol(j) for j in data.dimlabels])
+        variable_symbols = axis_names & all_symbols
+        parameter_symbols = all_symbols - variable_symbols
+        variable_symbols = tuple(variable_symbols)
+        variable_names = tuple([str(j) for j in variable_symbols])
+        parameter_symbols = tuple(parameter_symbols)
+        parameter_names = tuple([str(j) for j in parameter_symbols])
+        logger.debug(
+            strm(
+                "all symbols are",
+                all_symbols,
+                "axis names are",
+                axis_names,
+                "variable names are",
+                variable_names,
+                "parameter names are",
+                parameter_names,
+            )
         )
-    )
-    # }}}
-    pars = Parameters()
-    for this_name in parameter_names:
-        kwargs = {}
-        if this_name in guesses.keys():
-            logger.info(strm("applying bounds for", this_name))
-            kwargs.update(guesses[str(this_name)])
-        pars.add(this_name, **kwargs)
-    for j in pars:
-        logger.info(strm("fit param ---", j))
-    logger.info(strm(pars))
-    fn = lambdify(
-        variable_symbols + parameter_symbols,
-        expr,
-        modules=[{"ImmutableMatrix": np.ndarray}, "numpy", "scipy"],
-    )
-    return pars, fn
-def residual(pars, x, data=None):
-    "calculate the residual OR if data is None, return fake data"
-    shift = pars["shift"]
-    if abs(shift) > pi / 2:
-        shift = shift - sign(shift) * pi
-    model = myfunc(x, **pars.valuesdict())
-    if data is None:
-        return model
-    return model - data
+        # }}}
+        pars = Parameters()
+        for this_name in parameter_names:
+            kwargs = {}
+            if this_name in guesses.keys():
+                logger.debug(strm("applying bounds for", this_name))
+                kwargs.update(guesses[str(this_name)])
+            pars.add(this_name, **kwargs)
+        for j in pars:
+            logger.info(strm("fit param ---", j))
+        logger.info(strm(pars))
+        self.fn = lambdify(
+            variable_symbols + parameter_symbols,
+            expr,
+            modules=[{"ImmutableMatrix": np.ndarray}, "numpy", "scipy"],
+        )
+        return pars
```

Comment:
the *only* thing I change after wrapping it in a class  is this -- see how I'm storing the function as an attribute of the class rather than returning it -- explained more below.